Lib/multiprocessing/shared_memory.py

```diff
@@ -1,228 +1,234 @@
+"""
 
-__all__ = [ 'SharedMemory', 'PosixSharedMemory', 'WindowsNamedSharedMemory',
-            'ShareableList', 'shareable_wrap',
-            'SharedMemoryServer', 'SharedMemoryManager', 'SharedMemoryTracker' ]
 
+__all__ = [ 'SharedMemory', 'ShareableList' ]
 
-from functools import reduce
+
+from functools import partial
 import mmap
-from .managers import DictProxy, SyncManager, Server
-from . import util
 import os
-import random
+import errno
 import struct
-import sys
-try:
-    from _posixshmem import _PosixSharedMemory, Error, ExistentialError, O_CREX
-except ImportError as ie:
-    if os.name != "nt":
-        # On Windows, posixshmem is not required to be available.
-        raise ie
-    else:
-        _PosixSharedMemory = object
-        class ExistentialError(BaseException): pass
-        class Error(BaseException): pass
-        O_CREX = -1
-
-
-class WindowsNamedSharedMemory:
-
-    def __init__(self, name, flags=None, mode=None, size=None, read_only=False):
-        if name is None:
-            name = f'wnsm_{os.getpid()}_{random.randrange(100000)}'
-
-        self._mmap = mmap.mmap(-1, size, tagname=name)
-        self.buf = memoryview(self._mmap)
-        self.name = name
-        self.size = size
-
-    def __repr__(self):
-        return f'{self.__class__.__name__}({self.name!r}, size={self.size})'
-
-    def close(self):
-        self.buf.release()
-        self._mmap.close()
+import secrets
 
-    def unlink(self):
-        """Windows ensures that destruction of the last reference to this
-        named shared memory block will result in the release of this memory."""
-        pass
+if os.name == "nt":
+    import _winapi
+    _USE_POSIX = False
+else:
+    import _posixshmem
+    _USE_POSIX = True
 
 
-class PosixSharedMemory(_PosixSharedMemory):
+_O_CREX = os.O_CREAT | os.O_EXCL
 
-    def __init__(self, name, flags=None, mode=None, size=None, read_only=False):
-        if name and (flags is None):
-            _PosixSharedMemory.__init__(self, name)
-        else:
-            if name is None:
-                name = f'psm_{os.getpid()}_{random.randrange(100000)}'
-            _PosixSharedMemory.__init__(self, name, flags=O_CREX, size=size)
+# FreeBSD (and perhaps other BSDs) limit names to 14 characters.
+_SHM_SAFE_NAME_LENGTH = 14
 
-        self._mmap = mmap.mmap(self.fd, self.size)
-        self.buf = memoryview(self._mmap)
+# Shared memory block name prefix
+if _USE_POSIX:
+    _SHM_NAME_PREFIX = 'psm_'
+else:
+    _SHM_NAME_PREFIX = 'wnsm_'
 
-    def __repr__(self):
-        return f'{self.__class__.__name__}({self.name!r}, size={self.size})'
 
-    def close(self):
-        self.buf.release()
-        self._mmap.close()
-        self.close_fd()
+def _make_filename():
+    "Create a random filename for the shared memory object."
+    # number of random bytes to use for name
+    nbytes = (_SHM_SAFE_NAME_LENGTH - len(_SHM_NAME_PREFIX)) // 2
+    assert nbytes >= 2, '_SHM_NAME_PREFIX too long'
```

Comment:
This assert seems conservative. Given a prefix of length 11 or 12 (assuming a 14 character limit), nbytes would be 1, and you could still make 256 valid filenames. Is that enough? Could that number of filenames be exhausted?

Reply:
The choice to have any prefix at all is largely down to a desire to have some sense of human-recognizable pattern to the auto-generated names but there is no need or motivation to make such a long prefix.  Users can specify their own choice of name for a shared memory block.  This auto-generated name mechanism is purely for convenience so that users do not need to do so themselves.